### PR TITLE
(CDAP-12646) fix run record fixer

### PR DESCRIPTION
**Pull request by chtyim**
- Perform micro batches of scanning from store to avoid tx timeout
- Removed all unnecessary point lookup of run records
  - Make ProgramRunId available to RunRecordMeta
  - Updated AppMetaStore deserialize RunRecordMeta with the ProgramRunId in it (which comes from the MDS key during scan, no extra lookup is needed).

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java [hunk 1]
```diff
@@ -17,10 +17,170 @@
...
+    this.runtimeService = runtimeService;
+    this.startTimeoutSecs = startTimeoutSecs;
+    this.txBatchSize = txBatchSize;
+  }
+
+  void fixRunRecords() {
+    Set<ProgramRunId> fixed = doFixRunRecords();
+
+    if (!fixed.isEmpty()) {
+      LOG.info("Corrected {} of run records in {} status but no actual program running.",
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: Can you improve this message? (The params are reversed and the grammar is odd). Perhaps:

"Correct {n} run records with status in {statuses} that have no actual running program. Such programs likely have crashed or were killed by external signal. "

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java [hunk 2]
```diff
@@ -17,10 +17,170 @@
...
+        LOG.trace("{} run records in {} state but are not actually running", runs.size());
+        if (runs.isEmpty()) {
+          break;
+        }
+
+        for (RunRecordMeta record : runs.values()) {
+          startTime = Math.max(startTime, RunIds.getTime(record.getPid(), TimeUnit.SECONDS));
+
+          ProgramRunId programRunId = record.getProgramRunId();
+          if (!fixedPrograms.contains(programRunId)) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java:111]: isn't this already filtered by the ```createFilter(fixedProgram)``` above?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java [hunk 3]
```diff
@@ -17,10 +17,170 @@
...
+            programStateWriter.error(
+              programRunId,
+              new Throwable("Marking run record for " + programRunId + " as failed since no running program found."));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: Can we have a specific exception class for this? Seems odd to use Throwable...

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java [hunk 4]
```diff
@@ -17,10 +17,170 @@
...
+      @Override
+      public boolean apply(RunRecordMeta record) {
+        ProgramRunId programRunId = record.getProgramRunId();
+
+        if (excludedIds.contains(programRunId)) {
+          return false;
+        }
+
+        // runs are marked as starting before they get into the runtimeService list.
+        // Therefore, add some buffer so that we don't incorrectly 'fix' runs that are actually starting
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: comment is confusing here, this is actually done at the end of the method. Maybe move it down there?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 5]
```diff
@@ -159,8 +160,17 @@ public AppMetadataStore(Table table, CConfiguration cConf, AtomicBoolean upgrade
   }
 
   @Override
-  protected <T> T deserialize(byte[] serialized, Type typeOfT) {
-    return GSON.fromJson(Bytes.toString(serialized), typeOfT);
+  protected <T> T deserialize(MDSKey key, byte[] serialized, Type typeOfT) {
+    if (RunRecordMeta.class.equals(typeOfT)) {
+      RunRecordMeta meta = GSON.fromJson(Bytes.toString(serialized), RunRecordMeta.class);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:165]: can you explain this? Does the Json not have the correct program/run ID?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/RunRecordCorrectorServiceTest.java [hunk 6]
```diff
@@ -52,94 +54,141 @@
...
+    expectedStates.put(workflowId, ProgramRunStatus.COMPLETED);
+    expectedStates.put(mrInWorkflowId, ProgramRunStatus.FAILED);
+
+    // Validates all expected states again
+    for (Map.Entry<ProgramRunId, ProgramRunStatus> entry : expectedStates.entrySet()) {
+      validateExpectedState(entry.getKey(), entry.getValue());
+    }
+  }
+
+  private void validateExpectedState(ProgramRunId programRunId, ProgramRunStatus expectedStatus) {
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/RunRecordCorrectorServiceTest.java:189]: this test case should be a lot faster now, right? 👍

**Review by anew — COMMENTED** [review]
mostly minor comments from me. 

I understand that the major change here is (apart from refactoring) that we now store the programRunId as part of the run record and therefore can avoid the point lookup for each record. Or did I miss anything?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/RunRecordCorrectorServiceTest.java (hunk 6 above)
Comment (chtyim) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/RunRecordCorrectorServiceTest.java:189]: yes and test more conditions.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 5 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:165]: The `ProgramRunId` is transient, hence is not serialized when stored into MDS (for backward compatibility and also not to duplicate what's in the row key already). Maybe we can reduce the temporary object creation by using a custom type adapter if necessary.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java (hunk 4 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: Haven't look at the message... it was copied.. I think it means don't fix record right away as the program running state may not be in the `ProgramRuntimeService` yet. I'll rephase the message better.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java:111]: yeah, it is the same Set. Moving it outside the loop.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: Sure. This was copied without much thought.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java (hunk 3 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: Sure.

**Comment by chtyim** [comment]
Thanks for the review @anew . Yes, the idea is to simply include the `ProgramRunId` in the `RunRecordMeta` during scan, such that we can remove the unnecessary expensive reserve lookup from RunId to ProgramId.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 5 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:165]: I think it's fine. If I had read this from bottom to top, I would have seen the changes down there first :)

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java [hunk 7]
```diff
@@ -76,8 +76,9 @@ void fixRunRecords() {
     Set<ProgramRunId> fixed = doFixRunRecords();
 
     if (!fixed.isEmpty()) {
-      LOG.info("Corrected {} of run records in {} status but no actual program running.",
-               NOT_STOPPED_STATUSES, fixed.size());
+      LOG.info("Correct {} run records with status in {} that have no actual running program. " +
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: Oops this needs to be "Corrected"

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java [hunk 8]
```diff
@@ -108,9 +109,10 @@ void fixRunRecords() {
 
           ProgramRunId programRunId = record.getProgramRunId();
           if (!fixedPrograms.contains(programRunId)) {
-            programStateWriter.error(
-              programRunId,
-              new Throwable("Marking run record for " + programRunId + " as failed since no running program found."));
+            programStateWriter.error(programRunId,
+                                     new ProgramRunAbortedException("Marking run record for "
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: should this use the same message as the warning?

**Comment by anew** [comment]
Only two more comments about messages that slipped my radar the first time.

**Comment by chtyim** [comment]
Build passed https://builds.cask.co/browse/CDAP-RUT1410

**Review by anew — APPROVED** [review]
LGTM

**Comment by chtyim** [comment]
Squashed and merging.

**Pull request merged**